SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityMesosStatusUpdateHandler.java

```diff
@@ -200,15 +205,35 @@ private void unsafeProcessStatusUpdate(Protos.TaskStatus status, SingularityTask
       return;
     }
 
+    final Optional<SingularityTask> task = taskManager.getTask(taskIdObj);
+
     if (status.getState() == TaskState.TASK_LOST) {
-      lostTasksMeter.mark();
-      if (configuration.getDisasterDetection().isEnabled()) {
-        taskLostReasons.add(status.getReason());
+      boolean isMesosFailure =
+          status.getReason() == Reason.REASON_INVALID_OFFERS
+          || status.getReason() == Reason.REASON_AGENT_REMOVED
+          || status.getReason() == Reason.REASON_AGENT_RESTARTED
+          || status.getReason() == Reason.REASON_AGENT_UNKNOWN
+          || status.getReason() == Reason.REASON_MASTER_DISCONNECTED
+          || status.getReason() == Reason.REASON_AGENT_DISCONNECTED;
+
+      RequestType requestType = task.isPresent() ? task.get().getTaskRequest().getRequest().getRequestType() : null;
+      boolean isRelaunchable =
+          requestType != null
+              && (requestType == RequestType.ON_DEMAND
+                  || requestType == RequestType.SCHEDULED
+                  || requestType == RequestType.RUN_ONCE);
+
+      if (isMesosFailure && isRelaunchable) {
+        LOG.info("Relaunching lost task {}", task);
+        relaunchTask(task.get());
+      } else {
+        lostTasksMeter.mark();
```

Comment:
we'll still want the lost task meter to fire regardless of failure type. The lost task metric allows us to do things like alert on a large wave of these (indicating that something is wrong with singularity. Similar with the disaster detection piece below

Reply:
Won't this still cause the alerts that the original ticket wanted to resolve then?

Reply:
There weren't alerts in the original issue. The original issue was that a failure that was mesos'/singularity's fault was counting towards the retry count for a scheduled task. i.e. it was given 2 attempts, both failed due to the invalid offers reason. So, even though the task wasn't the thing that failed, it wasn't retried again